src/Annotator.ts

```diff
@@ -0,0 +1,232 @@
+    if (failures > 0) {
+      conclusion = Conclusion.Failure;
+    } else if (warnings > 0 || notices > 0) {
+      conclusion = Conclusion.Neutral;
+    }
+
+    return conclusion;
+  }
+
+  private generateSummary(failures: number, warnings: number, notices: number): string {
+    const messages = [];
+    if (failures > 0) {
+      const noun = failures === 1 ? 'failure' : 'failures';
+      messages.push(`${failures} ${noun} found`);
+    }
+
+    if (warnings > 0) {
+      const noun = warnings === 1 ? 'warning' : 'warnings';
+      messages.push(`${warnings} ${noun} found`);
+    }
+
+    if (notices > 0) {
+      const noun = notices === 1 ? 'notice' : 'notices';
+      messages.push(`${notices} ${noun} found`);
+    }
+
+    return messages.join('\n');
+  }
+
+  private stats(annotations: Annotation[]): AnnotationsPerLevel {
+    const result = { failures: 0, warnings: 0, notices: 0 };
+
+    annotations.forEach((annotation: Annotation) => {
+      const level = annotation.annotation_level;
+      if (level === AnnotationLevel.Failure) {
+        result.failures += 1;
+      } else if (level === AnnotationLevel.Notice) {
+        result.notices += 1;
+      } else if (level === AnnotationLevel.Warning) {
+        result.warnings += 1;
+      }
+    });
+
+    return result;
+  }
+
+  private async createCheck(): Promise<number> {
+    const { owner, repo, ref } = this;
+    const {
+      data: { id: checkRunId },
+    } = await this.octokit.rest.checks.create({
+      owner,
+      repo,
+      name: Annotator.CHECK_TITLE,
+      head_sha: ref,
+      status: 'in_progress',
+    });
+
+    return checkRunId;
+  }
+
+  private async updateCheck(
+    checkRunId: number,
+    conclusion: string,
+    summary: string,
+    annotations: Annotation[]
+  ): Promise<void> {
+    const { owner, repo } = this;
+    await this.octokit.rest.checks.update({
+      owner,
+      repo,
+      check_run_id: checkRunId,
+      status: 'completed',
+      conclusion,
+      output: {
+        title: Annotator.CHECK_TITLE,
+        summary,
+        annotations,
+      },
+    });
+  }
+
+  private fileAndLineNumberFromLocation(location: string | undefined): LocationInfo | undefined {
+    if (!location) return;
+    const locationTokens = location.split(':');
+    const lastToken = locationTokens[locationTokens.length - 1];
+
+    // if there are no colons in the location or the last token is not all digits
+    // then we can't use this location to make an annotation in the source code
+    if (locationTokens.length < 2 || !/^\d+$/.test(lastToken)) return;
+
+    const lineNumber = Number(locationTokens.pop());
+    const path = locationTokens.join(':');
+    return { path, lineNumber };
+  }
+
+  public annotationFromFinding(finding: Finding): Annotation | undefined {
+    const locationInfo = finding.stack.map(this.fileAndLineNumberFromLocation, this);
+
+    let preferredLocation = locationInfo.find((info) => info && !path.isAbsolute(info.path));
```

Comment:
In CI, we often see Ruby gems bundled into the _vendor_ directory, and it's also possible to see _node_modules_, so we should probably exclude any code that's in these directories, even if it's a relative path.

We should also add a property to the Action that allows the user to specify additionl directories to consider "non-local".